storage_scrubber/src/find_large_objects.rs

```diff
@@ -45,53 +50,85 @@ pub async fn find_large_objects(
-            for obj in fetch_response.contents().iter().filter(|o| {
-                if let Some(obj_size) = o.size {
-                    min_size as i64 <= obj_size
-                } else {
-                    false
+        let tenant_ctr = tenant_ctr.clone();
+        let object_ctr = object_ctr.clone();
+        let s3_client = s3_client.clone();
+        async move {
+            let mut objects = Vec::new();
+            // We want the objects and not just common prefixes
+            tenant_root.delimiter.clear();
+            let mut continuation_token = None;
+            loop {
+                let fetch_response =
+                    list_objects_with_retries(&s3_client, &tenant_root, continuation_token.clone())
+                        .await
+                        .expect("couldn't list objects");
+                for obj in fetch_response.contents().iter().filter(|o| {
+                    if let Some(obj_size) = o.size {
+                        min_size as i64 <= obj_size
+                    } else {
+                        false
+                    }
+                }) {
+                    let key = obj.key().expect("couldn't get key").to_owned();
+                    let kind = LargeObjectKind::from_key(&key);
+                    if ignore_deltas && kind == LargeObjectKind::DeltaLayer {
+                        continue;
+                    }
+                    objects.push(LargeObject {
+                        key,
+                        size: obj.size.unwrap() as u64,
+                        kind,
+                    })
                 }
-            }) {
-                let key = obj.key().expect("couldn't get key").to_owned();
-                let kind = LargeObjectKind::from_key(&key);
-                if ignore_deltas && kind == LargeObjectKind::DeltaLayer {
-                    continue;
+                object_ctr.fetch_add(fetch_response.contents().len() as u64, Ordering::Relaxed);
+                match fetch_response.next_continuation_token {
+                    Some(new_token) => continuation_token = Some(new_token),
+                    None => break,
                 }
-                objects.push(LargeObject {
-                    key,
-                    size: obj.size.unwrap() as u64,
-                    kind,
-                })
-            }
-            object_ctr += fetch_response.contents().len() as u64;
-            match fetch_response.next_continuation_token {
-                Some(new_token) => continuation_token = Some(new_token),
-                None => break,
             }
+
+            tenant_ctr.fetch_add(1, Ordering::Relaxed);
+
+            Ok((tenant_shard_id, objects))
         }
+    });
+    let mut objects_stream = std::pin::pin!(objects_stream.try_buffer_unordered(concurrency));
+
+    //let mut objects_stream = objects_stream.flatten();
+
+    let mut objects = Vec::new();
+    while let Some(res) = objects_stream.next().await {
```

Comment:
does this really increase speed? every `next` waits until a tenant finishes before continuing.

I think to achieve true concurrency, we need either `tokio::spawn` to create new tasks, or `futures_unordered` to poll multiple futures in a future.

Reply:
sorry I didn't see try_buffer_unordered